packages/react/tooltip/src/createStateMachine.tsx

```diff
@@ -0,0 +1,135 @@
+const ASSIGN_ACTION_TYPE = 'machine.actions.assign' as const;
+
+const isProduction = process.env.NODE_ENV === 'production';
+
+type StateChart<State extends string, Event extends { type: string }, Context> = {
+  initial: State;
+  context: Context;
+  states: {
+    [stateValue in State]: {
+      on?: {
+        [eventType in Event['type']]?: Transition<State, Event, Context>;
+      };
+      entry?: Array<Action<Event, Context>>;
+      exit?: Array<Action<Event, Context>>;
+    };
+  };
+};
+
+type Transition<State extends string, Event, Context> = {
+  target: State;
+  actions?: Array<Action<Event, Context>>;
+};
+
+type Action<Event, Context> = ActionFunction<Event, Context> | AssignAction<Event, Context>;
+
+type ActionFunction<Event, Context> = (
+  event: Event,
+  context: Context,
+  send: (event: Event) => void
+) => void;
+
+function createStateMachine<State extends string, Event extends { type: string }, Context>(
+  stateChart: StateChart<State, Event, Context>,
+  { debug = false, warnOnUnknownTransitions = !isProduction } = {}
+) {
+  let PREVIOUS_STATE: State | undefined;
+  let CURRENT_STATE = stateChart.initial;
+  let PREVIOUS_CONTEXT: Context | undefined;
+  let CURRENT_CONTEXT = stateChart.context;
+
+  type CallbackFn = (args: { state: State; previousState?: State; context: Context }) => void;
+  const subscriptions: Array<CallbackFn> = [];
+
+  function subscribe(callback: CallbackFn) {
+    subscriptions.push(callback);
+    return () => {
+      subscriptions.splice(subscriptions.indexOf(callback), 1);
+    };
+  }
+
+  function notify() {
+    subscriptions.forEach((callback) =>
+      callback({
+        state: CURRENT_STATE,
+        previousState: PREVIOUS_STATE,
+        context: CURRENT_CONTEXT,
+      })
+    );
+  }
+
+  const send = (event: Event) => {
+    const stateDefinition = stateChart.states[CURRENT_STATE];
+    const type: Event['type'] | undefined = event.type;
+    const transition: Transition<State, Event, Context> | undefined = stateDefinition.on?.[type];
+
+    if (transition === undefined) {
+      if (warnOnUnknownTransitions) {
+        console.warn(
+          `From state: "${CURRENT_STATE}", event "${type}" has no transition to any state`
+        );
+      }
+    } else {
+      PREVIOUS_STATE = CURRENT_STATE;
```

Comment:
I know you haven't changed this but it really confused me seeing uppercase variables being reassigned like this. I would usually expect all uppercase to be a constant.

Reply:
Good point. Is it ok with you if I make this change once we've merged the stack of PRs to make conflict handling easier?

Reply:
sure no prob 👍